src/tools/run-make-support/src/command.rs

```diff
@@ -151,6 +151,53 @@ impl Command {
         self
     }
 
+    /// Set an auxiliary stream passed to the process, besides the stdio streams.
+    ///
+    /// # Safety
+    ///
+    /// Use with caution! Ideally, only set one aux fd; if there are multiple, their old `fd` may
+    /// overlap with another's `new_fd`, and may break. The caller must make sure this is not the
+    /// case.
+    #[cfg(unix)]
+    pub unsafe fn set_aux_fd<F: Into<std::os::fd::OwnedFd>>(
+        &mut self,
+        new_fd: std::os::fd::RawFd,
+        fd: F,
+    ) -> &mut Self {
+        // NOTE: If more than 1 auxiliary file descriptor is needed, this function should be
+        // rewritten.
+
+        use std::os::fd::AsRawFd;
+        use std::os::unix::process::CommandExt;
+
+        let cvt = |x| if x == -1 { Err(std::io::Error::last_os_error()) } else { Ok(()) };
+
+        let fd = fd.into();
+        if fd.as_raw_fd() == new_fd {
+            // If the new file descriptor is already the same as fd, just turn off `FD_CLOEXEC`.
+
+            // SAFETY(io-safety): `fd` is already owned.
+            cvt(unsafe { libc::fcntl(fd.as_raw_fd(), libc::F_SETFD, 0) })
+                .expect("disabling CLOEXEC failed");
+            // The `pre_exec` function should be unconditionally set, since it captures `fd`, and
+            // this ensures that it stays open until the fork.
```

Comment:
Wouldn't it still be closed once pre_exec returns? I think you need to `mem::forget` it explicitly.
Personally I would put `fd` in a `ManuallyDrop` at the beginning of the function and only work with its `RawFd` from that point on, that seems less prone to mistakes.

Reply:
Switched to a `ManuallyDrop`, and used its `RawFd` afterwards.